### ISSUE TITLE
Verifier error when putting pointer in a tuple (pointer arithmetic with >>= operator prohibited)

### ISSUE THREAD

**Issue by ajor** [bpftrace/bpftrace#3479]
This seems to happen when putting any BTF pointer type in a tuple prior to printing.

Example 1:
```
kretfunc:__skb_recv_udp
{
  print((1, retval)); // retval is a "struct sk_buff *"
}
```
Example 2:
```
kfunc:__x64_sys_read
{
  print((1, args.regs)) // args.regs is a "const struct pt_regs *"
}
```

Verifier error from example 1:
```
Kernel error log: 
0: R1=ctx() R10=fp0
;  @ bpftrace.bpf.o:0
0: (79) r7 = *(u64 *)(r1 +32)
func '__skb_recv_udp' arg4 has btf_id 654 type STRUCT 'sk_buff'
1: R1=ctx() R7_w=ptr_sk_buff()
1: (b7) r6 = 0                        ; R6_w=0
2: (63) *(u32 *)(r10 -8) = r6         ; R6_w=0 R10=fp0 fp-8=????0
3: (bf) r2 = r10                      ; R2_w=fp0 R10=fp0
4: (07) r2 += -8                      ; R2_w=fp-8
5: (18) r1 = 0xffff888b82348300       ; R1_w=map_ptr(map=fmt_string_args,ks=4,vs=32)
7: (85) call bpf_map_lookup_elem#1    ; R0_w=map_value_or_null(id=1,map=fmt_string_args,ks=4,vs=32)
8: (55) if r0 != 0x0 goto pc+2 11: R0_w=map_value(map=fmt_string_args,ks=4,vs=32) R6_w=0 R7_w=ptr_sk_buff() R10=fp0 fp-8=????mmmm
11: (b7) r1 = 1                       ; R1_w=1
12: (73) *(u8 *)(r0 +16) = r1         ; R0_w=map_value(map=fmt_string_args,ks=4,vs=32) R1_w=1
13: (b7) r1 = 30007                   ; R1_w=30007
14: (7b) *(u64 *)(r0 +0) = r1         ; R0_w=map_value(map=fmt_string_args,ks=4,vs=32) R1_w=30007
15: (bf) r1 = r7                      ; R1_w=ptr_sk_buff() R7_w=ptr_sk_buff()
16: (77) r1 >>= 56
R1 pointer arithmetic with >>= operator prohibited
processed 16 insns (limit 1000000) max_states_per_insn 0 total_states 1 peak_states 1 mark_read 0
```
The verifier error from the second example is similar.

-----

bpftrace: latest master
Kernel: 6.9 with BPF backports.

**Comment by ZhangYet** [comment]
Hi @ajor I found I can run the example 1 without verifier error. My env is as blow:

```
sudo bpftrace --info
Alias tip: _ bpftrace --info
System
  OS: Linux 6.10.10-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 12 Sep 2024 17:21:02 +0000
  Arch: x86_64

Build
  version: v0.21.2
  LLVM: 18.1.8
  unsafe probe: no
  bfd: yes
  liblldb (DWARF support): no
  libsystemd (systemd notify support): no

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: yes
  get_boot_ns: yes
  dpath: yes
  skboutput: yes
  get_tai_ns: yes
  get_func_ip: yes
  jiffies64: yes
  for_each_map_elem: yes

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  module btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes
  ringbuf: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  kprobe_multi: yes
  uprobe_multi: yes
  raw_tp_special: yes
  iter: yes
```
Could you tell me your kernel tag? Let me find a way to reproduce it.

**Comment by viktormalik** [comment]
Same here, both the above examples work without problems on my kernel:
```
# bpftrace --info
System
  OS: Linux 6.10.5-200.fc40.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Aug 14 15:49:44 UTC 2024
  Arch: x86_64

Build
  version: v0.21.0-153-g6782
  LLVM: 18.1.6
[...]
```

**Comment by ajor** [comment]
> Could you tell me your kernel tag? Let me find a way to reproduce it.

I run a custom kernel with a bunch of BPF features backported so it's hard to compare, unfortunately. But thanks to the two of you pointing out that it does work on your kernels, I dug into it a bit more...

Turns out that it was a bug for a short period of time, introduced and fixed between the two versions of bpftrace you guys are running! (I musn't have been on latest master myself either)

Bug introduced in: e5167acba9b833adbb11ab8225fe18bba5bb7746
Bug fixed in: 1517d1cc9b7436069412d764ae32d34c9e58afd7

cc @danobi @ryantimwilson The bug is fixed now, but it seems that this scratch map work is having a wider impact than predicted - heads up!

**Comment by danobi** [comment]
Ack
